nemo/collections/tts/modules/transformer_dec24.py

```diff
@@ -0,0 +1,691 @@
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#############################################################################
+import math
+
+import torch
+import torch.nn.functional as F
+from flash_attn import flash_attn_varlen_kvpacked_func, flash_attn_varlen_qkvpacked_func
+from flash_attn.layers.rotary import RotaryEmbedding
+from torch import nn
+from torch.cuda import amp
+from torch.nn.utils.rnn import pad_sequence
+
+
+# alibi helper functions (start)
+def get_relative_positions(seq_len: int, device, symmetric: bool = True) -> torch.tensor:
+    x = torch.arange(seq_len)[None, :]
+    y = torch.arange(seq_len)[:, None]
+    rel_pos = (x - y).to(device)
+    if symmetric:
+        rel_pos = -rel_pos.abs()
+    return rel_pos
+
+
+def get_alibi_slope(num_heads):
+    x = (2**8) ** (1 / num_heads)
+    return torch.tensor([1 / x ** (i + 1) for i in range(num_heads)]).unsqueeze(-1).unsqueeze(-1)
+
+
+# alibi helper functions (end)
+
+
+class ConvNorm(torch.nn.Module):
+    def __init__(
+        self,
+        in_channels,
+        out_channels,
+        kernel_size=1,
+        stride=1,
+        padding=None,
+        dilation=1,
+        bias=True,
+        w_init_gain='gpt2',
+        is_causal=False,
+        norm_name='',
+    ):
+        super(ConvNorm, self).__init__()
+
+        padding = 0 if is_causal else padding
+        if padding is None:
+            assert kernel_size % 2 == 1
+            padding = int(dilation * (kernel_size - 1) / 2)
+
+        self.is_causal = is_causal
+        self.kernel_size = kernel_size
+        self.dilation = dilation
+
+        self.conv = torch.nn.Conv1d(
+            in_channels,
+            out_channels,
+            kernel_size=kernel_size,
+            stride=stride,
+            padding=padding,
+            dilation=dilation,
+            bias=bias,
+        )
+
+        if w_init_gain == 'gpt2':
+            torch.nn.init.normal_(self.conv.weight, mean=0.0, std=0.02)
+        else:
+            torch.nn.init.xavier_uniform_(self.conv.weight, gain=torch.nn.init.calculate_gain(w_init_gain))
+
+        self.norm_name = norm_name
+        if self.norm_name == 'weightnorm':
+            self.conv = nn.utils.weight_norm(self.conv)
+        elif self.norm_name == 'spectralnorm':
+            self.conv = nn.utils.spectral_norm(self.conv)
+
+    def forward(self, signal):
+        if self.is_causal:
+            padding = (int((self.kernel_size - 1) * (self.dilation)), 0)
+            signal = torch.nn.functional.pad(signal, padding)
+        conv_signal = self.conv(signal)
+        return conv_signal
+
+
+class LayerNorm(nn.Module):
```

Comment:
Is there a reason to use this instead of torch.nn.LayerNorm now?

Reply:
From what I understand, it is uses the same underlying function F.layer_norm, but always computes in fp32 to avoid numerical issues, which is why it is separate.